Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py

```diff
@@ -0,0 +1,393 @@
+        delete_processed: Whether to delete processed mails
+        limit: The maximum number of incidents to fetch each time
+        save_file: Whether to save the .eml file of the incident's mail
+
+    Returns:
+        next_run: This will be last_run in the next fetch-incidents
+        incidents: Incidents that will be created in Demisto
+    """
+    # Get the last fetch time, if exists
+    last_fetch = last_run.get('last_fetch')
+
+    # Handle first time fetch
+    if last_fetch is None:
+        latest_created_time = parse(f'{first_fetch_time} UTC')
+    else:
+        latest_created_time = datetime.fromisoformat(last_fetch)
+    mails_fetched, messages = fetch_mails(client=client,
+                                          include_raw_body=include_raw_body,
+                                          first_fetch_time=latest_created_time,
+                                          limit=limit,
+                                          permitted_from_addresses=permitted_from_addresses,
+                                          permitted_from_domains=permitted_from_domains,
+                                          save_file=save_file)
+    if mails_fetched:
+        latest_created_time = max(mails_fetched, key=lambda x: x.date).date
+    incidents = [mail.convert_to_incident() for mail in mails_fetched]
+    next_run = {'last_fetch': latest_created_time.isoformat()}
+    if delete_processed:
+        client.delete_messages(messages)
+    return next_run, incidents
+
+
+def fetch_mails(client: IMAPClient,
+                first_fetch_time: datetime = None,
+                permitted_from_addresses: str = '',
+                permitted_from_domains: str = '',
+                include_raw_body: bool = False,
+                limit: int = -1,
+                save_file: bool = False,
+                message_id: int = None):
+    """
+    This function will fetch the mails from the imap servers.
+
+    Args:
+        client: IMAP client
+        first_fetch_time: Fetch all incidents since first_fetch_time
+        include_raw_body: Whether to include the raw body of the mail in the incident's body
+        permitted_from_addresses: A string representation of list of mail addresses to fetch from
+        permitted_from_domains: A string representation list of domains to fetch from
+        limit: The maximum number of incidents to fetch each time, if the value is -1 all
+               mails will be fetched (used with list-messages command)
+        save_file: Whether to save the .eml file of the incident's mail
+        message_id: A unique message ID with which a specific mail can be fetched
+
+    Returns:
+        mails_fetched: A list of Email objects
+        messages: A list of the ids of the messages fetched
+    """
+    if message_id:
+        messages = [message_id]
+    else:
+        messages_query = generate_search_query(first_fetch_time,  # type: ignore[arg-type]
```

Comment:
we can avoid ignoring mypy validation here by handling better `first_fetch_time` type